timesync/ticker.go

```diff
@@ -0,0 +1,178 @@
+package timesync
+
+import (
+	"errors"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"sync"
+	"time"
+)
+
+// subs implements a lock-protected Subscribe-Unsubscribe structure
+// note: to access internal fields a lock must be obtained
+type subs struct {
+	subscribers map[LayerTimer]struct{} // map subscribers by channel
+	m           sync.Mutex
+}
+
+func newSubs() *subs {
+	return &subs{
+		subscribers: make(map[LayerTimer]struct{}),
+		m:           sync.Mutex{},
+	}
+}
+
+func (s *subs) Subscribe() LayerTimer {
+	ch := make(LayerTimer)
+	s.m.Lock()
+	s.subscribers[ch] = struct{}{}
+	s.m.Unlock()
+	log.Info("subscribed to channel")
+	return ch
+}
+
+func (s *subs) Unsubscribe(ch LayerTimer) {
+	s.m.Lock()
+	delete(s.subscribers, ch)
+	s.m.Unlock()
+}
+
+// LayerTimer is a channel of LayerIDs
+// Subscribers will receive the ticked layer through such channel
+type LayerTimer chan types.LayerID
+
+// LayerConverter provides conversions from time to layer and vice versa
+type LayerConverter interface {
+	TimeToLayer(time.Time) types.LayerID
+	LayerToTime(types.LayerID) time.Time
+}
+
+type Ticker struct {
+	*subs                 // the sub-unsub provider
+	clock           Clock // provides the time
+	started         bool
+	lastTickedLayer types.LayerID  // track last ticked layer
+	conv            LayerConverter // layer conversions provider
+	layerChannels   map[types.LayerID]chan struct{}
+	log             log.Log
+}
+
+func NewTicker(c Clock, lc LayerConverter) *Ticker {
+	return &Ticker{
+		subs:            newSubs(),
+		lastTickedLayer: lc.TimeToLayer(c.Now()),
+		clock:           c,
+		conv:            lc,
+		layerChannels:   make(map[types.LayerID]chan struct{}),
+		log:             log.NewDefault("ticker"),
+	}
+}
+
+var (
+	errNotStarted     = errors.New("ticker is not started")
+	errNotMonotonic   = errors.New("tried to tick a previously ticked layer")
+	errMissedTicks    = errors.New("missed ticks for one or more subscribers")
+	errMissedTickTime = errors.New("missed tick time by more than the allowed threshold")
+)
+
+const sendTickThreshold = 500 * time.Millisecond // allow up to sendTickThreshold
+
+// Notify notifies all the subscribers with the current layer
+// if the tick time has passed notify is skipped and errMissedTickTime is returned
+// notify may be skipped also for non-monotonic tick
+// if some of the subscribers where not listening, they are skipped. In that case, errMissedTicks is returned along the number of subscribers not listening
+func (t *Ticker) Notify() (int, error) {
+	if !t.started {
+		return 0, errNotStarted
+	}
+
+	t.m.Lock()
+
+	layer := t.conv.TimeToLayer(t.clock.Now())
+	// close prev layers
+	for l := t.lastTickedLayer + 1; l <= layer; l++ {
+		if layerChan, found := t.layerChannels[l]; found {
+			close(layerChan)
+			delete(t.layerChannels, l)
+		}
+	}
```

Comment:
This loop is new and should be tested.

There should be a test that simulates skipped ticks and ensures that if someone awaits those layers they are notified.

Reply:
Added